elastic/kibana
# Synthetics - project monitors with private locations

## Summary

Relates to https://github.com/elastic/uptime/issues/475. This PR builds on https://github.com/elastic/kibana/pull/135782. This PR should be merged after.

Adds support for private location support for project monitors, enabling the ability for private locations to be specified by name via the synthetics-agent.

Also adds in disclaimer text for users who do not have permissions. 

<img width="496" alt="Screen Shot 2022-07-20 at 6 03 13 PM" src="https://user-images.githubusercontent.com/11356435/180096356-13b22037-3ee4-4eeb-81f5-17e592b86d2e.png">

Testing
--
We can use https://github.com/elastic/elastic-package to test this PR

1. Pull Synthetics private location changes https://github.com/elastic/integrations/pull/3683/files and using cd into packages/synthetics, do `elastic-package clean` then `elastic-package build` and this will create conflict with registry running in docker desktop, stop that and run this command
2. From `packages/synthetics` Start elastic stack `elastic-package stack up --version 8.4.0-SNAPSHOT -v`
3. Once stack is up, stop kibana using docker desktop management
4. Add `xpack.fleet.registryUrl: https://localhost:8080` to `kibana.dev.yml`
5. Run `eval "$(elastic-package stack shellinit)"` then `export NODE_EXTRA_CA_CERTS=${ELASTIC_PACKAGE_CA_CERT}`. Then pull this PR and start kibana
6. Create a user with Uptime all, Fleet all, and Integrations all permissions
7.  Navigate to Uptime with this user and generate an api key
8. Click Manage Locations and create a location tied to the `elastic-agent (elastic-package)` agent policy.
9. Pull down https://github.com/elastic/synthetics/pull/559 and run `npm run build`.
10. Add the following to one of the example journeys in todos ahead of any steps
```
monitor.use({
    schedule: 3,
    privateLocations: ["YOUR_LOCATION_NAME"],
  })
```
11. In the root directory of the synthetics project directory `cd/examples/todos`.
12. node ../../dist/cli.js push  --url YOUR_KIBANA_URL --project test-project --auth YOUR_API_KEY --schedule 3 --locations us_east
13. Navigate back to Uptime. Ensure your monitor appears in the monitor list
14. Navigate to `app/fleet/policies/elastic-agent-managed-ep`. Ensure a Synthetics policy was created
15. Navigate to Uptime overview. Ensure your data is indexed properly
16. Remove the `privateLocations` key for your monitor and repush. 
17. Navigate back to Uptime. Ensure the location was removed
18. Navigate to `app/fleet/policies/elastic-agent-managed-ep`. Ensure the Synthetics policy was deleted
19. Run node ../../dist/cli.js push  --url YOUR_KIBANA_URL --project test-project --auth YOUR_API_KEY --schedule 3 --locations us_east --privateLocations "YOUR_LOCATION_NAME"
20. Navigate back to Uptime monitor list. Ensure the location was readded.
21. Navigate to `app/fleet/policies/elastic-agent-managed-ep`. Ensure a new Synthetics policy was created.
